### PR TITLE
fix: Adds check for .aab format before querying apkDetails

### LINKED ISSUE #2188: Bad Request error on getApkDetails when using .aab instead of .apk
**Describe the bug**
I have a GitLab pipeline with Android Instrumentation tests run by Flank.
Starting from Flank version 21.07.0 I have this error in my GitLab pipeline:
<pre>
400 Bad Request
POST https://testing.googleapis.com/v1/applicationDetailService/getApkDetails
{
  "code": 400,
  "errors": [
    {
      "domain": "global",
      "message": "Request contains an invalid argument.",
      "reason": "badRequest"
    }
  ],
  "message": "Request contains an invalid argument.",
  "status": "INVALID_ARGUMENT"
}
</pre>

<details>
<summary>Full stack trace:</summary>
<pre>
com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
POST https://testing.googleapis.com/v1/applicationDetailService/getApkDetails
{
  "code": 400,
  "errors": [
    {
      "domain": "global",
      "message": "Request contains an invalid argument.",
      "reason": "badRequest"
    }
  ],
  "message": "Request contains an invalid argument.",
  "status": "INVALID_ARGUMENT"
}
 at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
 at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
 at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
 at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:428)
 at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1111)
 at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:514)
 at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:455)
 at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:565)
 at ftl.client.google.AppDetailsKt.getAndroidAppDetails(AppDetails.kt:7)
 at ftl.run.platform.RunAndroidTestsKt.reportPackageName(RunAndroidTests.kt:101)
 at ftl.run.platform.RunAndroidTestsKt.access$reportPackageName(RunAndroidTests.kt:1)
 at ftl.run.platform.RunAndroidTestsKt$runAndroidTests$2.invokeSuspend(RunAndroidTests.kt:54)
 at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
 at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
 at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
 at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
 at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
 at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
 at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
 at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
 at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
 at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
 at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
 at ftl.domain.RunTestAndroidKt.invoke(RunTestAndroid.kt:66)
 at ftl.presentation.cli.firebase.test.android.AndroidRunCommand.run(AndroidRunCommand.kt:54)
 at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
 at picocli.CommandLine.access$1300(CommandLine.java:145)
 at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
 at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
 at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
 at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
 at picocli.CommandLine.execute(CommandLine.java:2078)
 at ftl.Main$main$1.invoke(Main.kt:12)
 at ftl.Main$main$1.invoke(Main.kt:10)
 at ftl.run.exception.ExceptionHandlerKt.withGlobalExceptionHandling(ExceptionHandler.kt:28)
 at ftl.run.exception.ExceptionHandlerKt.withGlobalExceptionHandling(ExceptionHandler.kt:17)
 at ftl.Main.main(Main.kt:10)
</pre>
</details>

After investigation I found out that `getApkDetails` expects `gcsPath` to `.apk` file to be sent, but Flank is supporting `.aab` as well and I'm using it.
There is a link to `getApkDetails` documentation - [getApkDetails](https://firebase.google.com/docs/test-lab/reference/testing/rest/v1/applicationDetailService/getApkDetails)
And there is a link to `FileReference` object documentation - [FileReference](https://firebase.google.com/docs/test-lab/reference/testing/rest/v1/projects.testMatrices#FileReference). In example we can see `.apk` file.

**To Reproduce**
Preconditions:
- Flank version 21.07.0 or higher
- `app` attribute in `flank.yml` contains path to `.aab` file

<details>
<summary>flank.yml example</summary>
<pre>
gcloud:
  app: ${APP_AAB}
  test: ${ANDROID_TEST_APK}
  use-orchestrator: true
  auto-google-login: true
  record-video: true
  timeout: 30m
  type: instrumentation
  num-flaky-test-attempts: 2
  environment-variables:
    clearPackageData: true
  device:
    # Google Pixel 5e (physical)
    - model: redfin
      version: 30
      locale: en_GB
      orientation: portrait
flank:
  shard-time: 180
  default-test-time: 15
  max-test-shards: 20
  run-timeout: 60m
</pre>

`APP_AAB` contains local path to `.aab`
`ANDROID_TEST_APK` contains local path to Android Test `.apk`
</details>

1. Run `java -jar ./flank.jar firebase test android run`
2. See error
3. Test Matrix is not created, tests are not running

**Expected behavior**
Error is not appearing
Test Matrix is created, tests are running

**Details**
It wasn't checked on the latest snapshot, but due to changes between it and version 21.11.0, I believe that the issue will be reproducible there.

Issue is reproducible on versions:
- 21.07.0
- 21.07.1
- 21.08.1
- 21.11.0

Issue is not reproducible on versions:
- 21.05.0
- 21.06.1

Other versions weren't tested.

If I replace my `.aab` file by `.apk` file in the `flank.yml` - issue disappears, so looks like it is a root cause.

### PR DESCRIPTION
**Pull request by Xlopec**
Fixes #2188

This PR adds checks for packaging format before querying for APK details. It fixes the following [issue #2188](https://github.com/Flank/flank/issues/2188)

## Checklist

- [] Documented
- [x] Unit tested
- [] Integration tests updated

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt [hunk 1]
```diff
@@ -13,18 +13,8 @@ import ftl.args.shardsFilePath
...
-import ftl.run.model.RoboTestContext
-import ftl.run.model.SanityRoboTestContext
-import ftl.run.model.TestResult
-import ftl.run.platform.android.asMatrixTestShards
-import ftl.run.platform.android.createAndroidTestConfig
-import ftl.run.platform.android.createAndroidTestContexts
-import ftl.run.platform.android.createAndroidTestMatrixType
-import ftl.run.platform.android.upload
+import ftl.run.model.*
+import ftl.run.platform.android.*
```
Comment (bootstraponline) [inline test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt:17]: per the lint rules, we avoid wildcard imports.

File: test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt [hunk 2]
```diff
@@ -50,7 +40,10 @@ internal suspend fun AndroidArgs.runAndroidTests(): TestResult = coroutineScope
                 ignoredTestsShardChunks += context.ignoredTestCases
                 allTestShardChunks += context.shards
             }
-            context.reportPackageName()
+
+            if (args.appApk?.endsWith(".aab") != true) {
```
Comment (bootstraponline) [inline test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt:44]: I think we should only report if it ends with `.apk` as google may add additional extensions in the future

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt [hunk 3]
```diff
@@ -31,17 +33,12 @@ import ftl.test.util.TestHelper.getPath
 import ftl.test.util.TestHelper.getThrowable
 import ftl.test.util.assertThrowsWithMessage
 import ftl.util.asFileReference
-import io.mockk.every
-import io.mockk.mockkObject
-import io.mockk.mockkStatic
-import io.mockk.unmockkAll
+import ftl.util.getMockedTestMatrix
+import ftl.util.mockTestMatrices
+import io.mockk.*
```
Comment (bootstraponline) [inline test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:38]: let's revert the wildcard imports

File: test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt [hunk 4]
```diff
@@ -1717,6 +1715,34 @@ AndroidArgs
             }
     }
 
+    @Test
+    fun `should send no package name to mixpanel for aab format`() {
```
Comment (bootstraponline) [inline test_runner/src/test/kotlin/ftl/args/AndroidArgsTest.kt:1719]: thank you for adding a test!

File: test_runner/src/test/kotlin/ftl/util/MatricesMocks.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    testExecutions = listOf(
+        TestExecution().apply {
+            resultStorage = ResultStorage().apply {
+                googleCloudStorage = GoogleCloudStorage().apply {
+                    gcsPath = "any/Path"
+                }
+            }
+        }
+    )
+}
```
Comment (bootstraponline) [inline test_runner/src/test/kotlin/ftl/util/MatricesMocks.kt:33]: let's add a newline to the end of the file to make git happy

File: test_runner/src/test/kotlin/ftl/run/TestRunnerTest.kt [hunk 6]
```diff
@@ -181,17 +180,4 @@ class TestRunnerTest {
...
-                resultStorage = ResultStorage().apply {
-                    googleCloudStorage = GoogleCloudStorage().apply {
-                        gcsPath = "any/Path"
-                    }
-                }
-            }
-        )
-    }
-}
+}
```
Comment (bootstraponline) [inline test_runner/src/test/kotlin/ftl/run/TestRunnerTest.kt:183]: add newline

**Review by bootstraponline — CHANGES_REQUESTED** [review]
Thanks for submitting a fix for this issue! I've left some comments.

File: test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt (hunk 2 above)
Comment (Xlopec) [reply test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt:44]: Ok, what about `null` case, should we reportPackageName in this case? I guess we will get FileReference.remote == "" in this case

**Comment by bootstraponline** [comment]
@Xlopec it looks like the code is failing lint check. Details in the GitHub actions build jobs.

> lintKotlinMain sources failed lint check

File: test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt (hunk 2 above)
Comment (bootstraponline) [reply test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt:44]: What do you think about this approach?

If the app path doesn't end in `.apk`, we'll report an empty string. App path can't be null.

```diff
diff --git a/test_runner/src/main/kotlin/ftl/client/google/AppDetails.kt b/test_runner/src/main/kotlin/ftl/client/google/AppDetails.kt
index 0a806b70..118b74cd 100644
--- a/test_runner/src/main/kotlin/ftl/client/google/AppDetails.kt
+++ b/test_runner/src/main/kotlin/ftl/client/google/AppDetails.kt
@@ -3,10 +3,14 @@ package ftl.client.google
 import com.google.testing.model.FileReference
 import ftl.http.executeWithRetry

-fun getAndroidAppDetails(gcsAppPath: String): String =
-    GcTesting.get
+fun getAndroidAppDetails(gcsAppPath: String): String {
+    // getApkDetails errors when sent non-apk files such as aab
+    if (gcsAppPath.trim().lowercase().endsWith(".apk").not()) return ""
+
+    return GcTesting.get
         .ApplicationDetailService()
         .getApkDetails(FileReference().apply { gcsPath = gcsAppPath })
         .apply { requestHeaders.set("X-Server-Timeout", 1800) } // 30 min
         .executeWithRetry()
         ?.apkDetail?.apkManifest?.packageName?.toString().orEmpty()
+}
```

Comment (Xlopec) [reply test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt]: Looks good to me

**Comment by bootstraponline** [comment]
@Xlopec If you could sign the CLA, then the builds should turn green.

image
image

**Review by bootstraponline — APPROVED** [review]

**Comment by Xlopec** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Xlopec** [comment]
recheck

**Comment by bootstraponline** [comment]
@Xlopec It looks like the test you added is failing in CI.

**Comment by bootstraponline** [comment]
image

**Pull request merged**